src/main/java/teammates/storage/search/FeedbackResponseCommentSearchDocument.java

```diff
@@ -431,4 +435,133 @@ private static boolean isNameVisibleToInstructor(
+                    filteredResultsSize -= sizeOfCommentList;
+                    // TODO: also need to decrease the size for (fr)CommentSearchResults|studentSearchResults
+                    frCommentSearchResults.comments.remove(response.getId());
+                    fr.remove();
+                }
+            }
+        }
+
+        Set<String> emailList = frCommentSearchResults.instructorEmails;
+        Iterator<Entry<String, List<FeedbackQuestionAttributes>>> iterQn =
+                frCommentSearchResults.questions.entrySet().iterator();
+        while (iterQn.hasNext()) {
+            String fsName = iterQn.next().getKey();
+            List<FeedbackQuestionAttributes> questionList = frCommentSearchResults.questions.get(fsName);
+
+            for (int i = questionList.size() - 1; i >= 0; i--) {
+                FeedbackQuestionAttributes question = questionList.get(i);
+                List<FeedbackResponseAttributes> responseList = frCommentSearchResults.responses.get(question.getId());
+
+                for (int j = responseList.size() - 1; j >= 0; j--) {
+                    FeedbackResponseAttributes response = responseList.get(j);
+                    List<FeedbackResponseCommentAttributes> commentList =
+                            frCommentSearchResults.comments.get(response.getId());
+
+                    for (int k = commentList.size() - 1; k >= 0; k--) {
+                        FeedbackResponseCommentAttributes comment = commentList.get(k);
+
+                        if (emailList.contains(comment.giverEmail)) {
+                            continue;
+                        }
+
+                        boolean isVisibilityFollowingFeedbackQuestion = comment.isVisibilityFollowingFeedbackQuestion;
+                        boolean isVisibleToGiver = isVisibilityFollowingFeedbackQuestion
+                                                 || comment.isVisibleTo(FeedbackParticipantType.GIVER);
+
+                        if (isVisibleToGiver && emailList.contains(response.giver)) {
+                            continue;
+                        }
+
+                        boolean isVisibleToReceiver = isVisibilityFollowingFeedbackQuestion
+                                                    ? question.isResponseVisibleTo(FeedbackParticipantType.RECEIVER)
+                                                    : comment.isVisibleTo(FeedbackParticipantType.RECEIVER);
+
+                        if (isVisibleToReceiver && emailList.contains(response.recipient)) {
+                            continue;
+                        }
+
+                        boolean isVisibleToInstructor = isVisibilityFollowingFeedbackQuestion
+                                                      ? question.isResponseVisibleTo(FeedbackParticipantType.INSTRUCTORS)
+                                                      : comment.isVisibleTo(FeedbackParticipantType.INSTRUCTORS);
```

Comment:
Might want to fix the custom indentation here too while you're transferring this over.

This method also has potential for refactoring (4 loops), but this likely shouldn't be done in this PR.